### PR TITLE
Handle undefined weight and bias in batch norm

### PR DESCRIPTION
**Pull request by asuhan**
Also allow inputs which aren't 4-D.

### CODE REVIEW

File: torch_xla/csrc/tensor.cpp [hunk 1]
```diff
@@ -265,7 +265,12 @@ XLATensor XLATensor::detach() const {
   XLATensor detached = XLATensor::clone(*this);
   detached.detach_();
   return detached;
-};
+}
+
+bool XLATensor::IsUndefined() const {
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor.cpp]: How do we end up creating tensors like these?
The bridge::CreateXlaTensor() checks if tensor is defined.
I don't like the have an undefined XLA tensor, having a Data member, with an ATEN tensor in it, which is undefined.
An undefined XLA tensor should be a tensor with Data which is nullptr.
This is why there is an is_null() API, and the default ctor is "defaulted" and public.

Comment (asuhan) [reply torch_xla/csrc/tensor.cpp]: We create them from undefined ATen tensors. Since we can already represent ATen tensors in the `Data`, I didn't want to add yet another quirk of having data as `nullptr` being undefined. The main issue I have with it is that null data should mean empty tensor, not undefined -- not being to distinguish the two is weird. Therefore, if we want to represent both empty and undefined tensors, we'll need a singleton Data to represent undefined.

**Comment by dlibenzi** [comment]
But there is already one way to create undefined tensors, and you changed
it.
An undefined tensor does not need to crash a bogus Data with an undefined
Aten Tensor in it.
As an undefined tensor does not need to carry any properties, we need not
to allocate a Data.
Can you point me where we are creating XLA tensors from undefined Aten ones?

**Comment by asuhan** [comment]
We create the XLA tensors from undefined ATen tensors in `AtenXlaType::batch_norm`, for example `weight` and `bias`.

I don't agree `Data` with an undefined ATen tensor in it is bogus. We merely wrap the undefine literal ATen tensor like we do for every other literal ATen tensor. PyTorch chooses to have a sentinel value to represent undefined tensors within that domain. We don't do anything besides reflecting that choice and provide access to it.

Also, we don't have a way to create *undefined* tensors. We have a way to create *empty* tensors. These two concepts are not the same; we can choose to throw away that distinction or add a singleton value for undefined `Data`. Either way we'll have to carry the burden of converting / representing undefined despite the fact that we already have the means to do that using the ATen sentinel.

I still prefer the current version, but the other options are ok too. Is your preference converting the undefined to `XLATensor`s with null data? We can do that.

File: torch_xla/csrc/tensor.cpp (hunk 1 above)
Comment (dlibenzi) [reply torch_xla/csrc/tensor.cpp]: But why do we want to have a Data for it?
You don't have to look far (besides an obvious unique_ptr/shared_ptr analogy), to see how "undefined tensors" are implemented.
Just look at at::Tensor default ctor, and you will notice it is exactly the same I am suggesting.
A data structure with an weak_ptr which is nullptr.

Comment (asuhan) [reply torch_xla/csrc/tensor.cpp]: The `impl_` in `at::Tensor` has `UndefinedTensorImpl` as a null type, which in turn only has one instance, its private `_singleton` field. `UndefinedTensorImpl` is also a `TensorImpl`. The closest we have to `TensorImpl` in `XLATensor` is `Data`. I don't fully understand your suggestion here, but we could introduce an `XLATensorImpl` to do similar things.

**Comment by dlibenzi** [comment]
The suggestion is to define an undefined XLA tensor as an XLATensor data
structure with a nullptr Data shared_ptr (like the default ctor for an
XLATensor does).

**Comment by asuhan** [comment]
Ok, I fail to see how this is different from saying empty and undefined tensors are the same for us. There's no way you could tell whether this tensor, with a nullptr `data_` in it, is empty or undefined. This is fine, but I'm not sure this is what you mean.

**Comment by dlibenzi** [comment]
> Ok, I fail to see how this is different from saying empty and undefined tensors are the same for us. There's no way you could tell whether this tensor, with a nullptr `data_` in it, is empty or undefined. This is fine, but I'm not sure this is what you mean.

The empty() series of tensors have a different meaning, if you mean "empty" in the ATEN empty() API way.
An undefined ATEN Tensor is an at::Tensor data structure which has a nullptr for the impl_ field.
Similarly, an undefined XLATensor will be an XLATensor constructed with the default ctor, which initializes the std::shared_ptr<Data> with nullptr.
Finally, an undefined ATEN/XLA Tensor will be an ATEN Tensor, with the impl_ field holding an XLA's tensor_impl, which hold an uninitialized XLATensor.

IOW:
```
AtenFromXlaTensor(XLATensor());
```

**Comment by asuhan** [comment]
Sounds good -- `ATen` will still represent undefined through `UndefinedTensorImpl`, no information is lost. I'll make the change.

**Comment by dlibenzi** [comment]
> Sounds good -- `ATen` will still represent undefined through `UndefinedTensorImpl`, no information is lost. I'll make the change.

To be clear ... I don't think we need an XLAUndefinedTensorImpl implementation for this.
When you do:
```
at:Tensor undefined;
```

You enter here:

https://github.com/pytorch/pytorch/blob/f084c129db176e148258b754dd97c4b9cce47326/aten/src/ATen/core/Tensor.h#L50

**Comment by dlibenzi** [comment]
And this is the defined() implementation:

https://github.com/pytorch/pytorch/blob/f084c129db176e148258b754dd97c4b9cce47326/aten/src/ATen/core/Tensor.h#L103

**Comment by asuhan** [comment]
Yes, I agree we don't. ATen provides the undefined for us, we just need to have an unitialized / empty `XLATensor`. It's asymmetrical, but `XLATensor` isn't meant to be a copycat of ATen anyway.

File: torch_xla/csrc/tensor.cpp (hunk 1 above)
Comment (asuhan) [reply torch_xla/csrc/tensor.cpp]: Done.

File: torch_xla/csrc/tensor_methods.cpp [hunk 2]
```diff
@@ -135,6 +135,24 @@ std::vector<xla::int64> CheckIntList(
...
+  xla::PrimitiveType input_element_type =
+      MakeXlaPrimitiveType(input.dtype(), &input.GetDevice());
+  return xla::ShapeUtil::MakeShape(input_element_type, {input.size(1)});
+}
+
+// Returns the IR for the given input or the provided default value broadcasted
+// to the default shape, if the input is undefined.
+ir::Value GetIrValueOrDefault(const XLATensor& input, at::Scalar default_value,
+                              const xla::Shape& default_shape) {
+  if (input.is_null()) {
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor_methods.cpp]: A ternary might not look that bad here.

Comment (asuhan) [reply torch_xla/csrc/tensor_methods.cpp]: Done.
